main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction2.scala

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright 2024 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.typer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.*
+import ca.uwaterloo.flix.language.ast.Type.JvmMember
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.phase.typer.ConstraintSolver2.Progress
+import ca.uwaterloo.flix.util.JvmUtils
+import ca.uwaterloo.flix.util.collection.ListMap
+import org.apache.commons.lang3.reflect.{ConstructorUtils, MethodUtils}
+
+import java.lang.reflect.{Constructor, Field, Method}
+
+object TypeReduction2 {
+
+  /**
+    * Performs various reduction rules on the given type.
+    */
+  def reduce(tpe0: Type)(implicit scope: Scope, tracker: Progress, renv0: RigidityEnv, eqenv: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef], flix: Flix): Type = tpe0 match {
```

Comment:
same comment about actual arguments vs. other stuff.

Reply:
eqenv -> eqenv0?

Reply:
I use `0` to mark when a variable can change in a recursive call. So actually it should be `renv` :)